TELCODOCS-1476 Hub-side templating (#69159)
openshift/openshift-docs: modules/ztp-specifying-nics-in-pgt-crs-with-hub-cluster-templates.adoc

:_mod-docs-content-type: PROCEDURE
[id="ztp-specifying-nics-in-pgt-crs-with-hub-cluster-templates_{context}"]
= Specifying group and site configuration in group PolicyGenTemplate CRs with hub templates

You can manage the configuration of fleets of clusters with `ConfigMap` CRs by using hub templates to populate the group and site values in the generated policies that get applied to the managed clusters.
Using hub templates in site `PolicyGenTemplate` (PGT) CRs means that you do not need to create a `PolicyGenTemplate` CR for each site.

> You can group the clusters in a fleet in various categories, depending on the use case, for example, hardware type or region. 

Reviewer: Suggestion- when "for example" is not the introduction of a sentence, I don't think it needs to be followed by a comma:
```suggestion
You can group the clusters in a fleet in various categories, depending on the use case, for example hardware type or region. 
```

> Each cluster should have a label corresponding to the group(s) that the cluster is in. 

Author: thanks @jeana-redhat I'e corrected that.

> If you manage the configuration values for each group in different `ConfigMap` CRs, then you require just one group `PolicyGenTemplate` CR to apply the changes to all the clusters in the group by using hub templates.

Reviewer: Per the [ISG word usage](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#j) entry on "just": "Omit just if the omission does not change the sentence meaning. Do not use just to mean only; use only instead."

If just here is used to mean only, I'd change it to say only:
```suggestion
If you manage the configuration values for each group in different `ConfigMap` CRs, then you require only one group `PolicyGenTemplate` CR to apply the changes to all the clusters in the group by using hub templates.
```

:_mod-docs-content-type: PROCEDURE
[id="ztp-specifying-nics-in-pgt-crs-with-hub-cluster-templates_{context}"]

> = Specifying group and site configuration in group PolicyGenTemplate CRs with hub cluster templates

Reviewer 2: ```suggestion
= Specifying group and site configuration in group PolicyGenTemplate CRs with hub templates
```

> = Specifying group and site configuration with group PolicyGenTemplate CRs with hub cluster templates

Reviewer 2: ```suggestion
= Specifying group and site configuration in group PolicyGenTemplate CRs with hub cluster templates
```

> You can manage the configuration of fleets of clusters with `ConfigMap` CRs and use hub cluster templates to populate the group and site values in the generated policies that get applied to the cluster hosts.

Reviewer 2: ```suggestion
You can manage the configuration of fleets of clusters with `ConfigMap` CRs and by using hub cluster templates to populate the group and site values in the generated policies that get applied to the cluster hosts.
```

> You can manage the configuration of fleets of clusters with `ConfigMap` CRs by using hub templates to populate the group and site values in the generated policies that get applied to the cluster hosts.

Reviewer 2: ```suggestion
You can manage the configuration of fleets of clusters with `ConfigMap` CRs by using hub templates to populate the group and site values in the generated policies that get applied to the spoke clusters.
```

> Using hub templates in site `PolicyGenTemplate` (PGT) CRs means that you do not need to create single-site `PolicyGenTemplate` CRs for each site.

Reviewer 3: ```suggestion
Using hub templates in site `PolicyGenTemplate` (PGT) CRs means that you do not need to create a `PolicyGenTemplate` CR for each site.
```

> Using hub cluster templates in site `PolicyGenTemplate` (PGT) CRs means that you do not need to create single-site `PolicyGenTemplate` CRs for each site.

Reviewer 2: ```suggestion
Using hub templates in site `PolicyGenTemplate` (PGT) CRs means that you do not need to create single-site `PolicyGenTemplate` CRs for each site.
```

You can group the clusters in a fleet in various categories, depending on the use case, for example, hardware type or region. 

> Each cluster should have a label corresponding to the group(s) the cluster is in. 

Reviewer 3: ```suggestion
Each cluster should have a label corresponding to the group(s) that the cluster is in. 
```

> If the configuration values for each group are kept in different `ConfigMap` CRs, you require just one group `PolicyGenTemplate` CR using hub cluster templates, and the cluster’s name and labels, to retrieve the values for each cluster.

Reviewer 2: ```suggestion
If the configuration values for each group are kept in different `ConfigMap` CRs, you require just one group `PolicyGenTemplate` CR using hub templates, and the cluster’s name and labels, to retrieve the values for each cluster.
```

> If the configuration values for each group are kept in different `ConfigMap` CRs, you require just one group `PolicyGenTemplate` CR using hub templates, and the cluster’s name and labels, to retrieve the values for each cluster.

Reviewer 3: ```suggestion
If you manage the configuration values for each group in different `ConfigMap` CRs, then, using hub templates, you require just one group `PolicyGenTemplate` CR to apply the configuration changes to to all the clusters in the group.
You use the cluster name and labels to retrieve the values for each cluster and apply the changes using hub templates in the `PolicyGenTemplate` YAML.
```

> 

Reviewer 3: ```suggestion
```

> The following example shows you how to use three `ConfigMap` CRs and one `PolicyGenTemplate` group CR to apply both site and group configuration to clusters grouped by hardware type and region.

Reviewer 2: ```suggestion
The following example shows you how to use three `ConfigMap` CRs and one group `PolicyGenTemplate` CR to apply both site and group configuration to clusters grouped by hardware type and region.
```

.Procedure

> . Create three `ConfigMap` CRs that contain the group and site configuration. 

Reviewer: Suggestion - since the following substeps are all parts of doing the action in this step, I think a colon might indication a stronger relationship. Up to you though.
```suggestion
. Create three `ConfigMap` CRs that contain the group and site configuration:
```

+
.. Create  a `ConfigMap` CR named `group-hardware-types-configmap` to hold the hardware-specific configuration. For example:
+
[source,yaml]
----

> ---

Reviewer: Just out of curiosity - are these three dashes in this yaml and the two following yaml files correct or necessary? They show up a bit weird in the rendered doc, but if they're needed then it is what it is.

Author: they're not needed so I've removed them. Thanks @skopacz1

  hardware-type-1-hugepages-count: "32"
----
<1> The `argocd.argoproj.io/sync-options` annotation is required only if the `ConfigMap` is larger than 1 MiB in size.

.. Create  a `ConfigMap` CR named `group-zones-configmap` to hold the regional configuration. For example:
+
[source,yaml]
----
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: group-zones-configmap
  namespace: ztp-group
data:
  # ClusterLogForwarder.yaml
  zone-1-cluster-log-fwd-outputs: "[{\"type\":\"kafka\", \"name\":\"kafka-open\", \"url\":\"tcp://10.46.55.190:9092/test\"}]"
  zone-1-cluster-log-fwd-pipelines: "[{\"inputRefs\":[\"audit\", \"infrastructure\"], \"labels\": {\"label1\": \"test1\", \"label2\": \"test2\", \"label3\": \"test3\", \"label4\": \"test4\"}, \"name\": \"all-to-default\", \"outputRefs\": [\"kafka-open\"]}]"
----

.. Create a `ConfigMap` CR named `site-data-configmap` to hold the site-specific configuration. For example:
+
[source,yaml]
----
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: site-data-configmap
  namespace: ztp-group
data:
  # SriovNetwork.yaml
  du-sno-1-zone-1-sriov-network-vlan-1: "140"
  du-sno-1-zone-1-sriov-network-vlan-2: "150"
----
+

> [NOTE]

Reviewer: This note seems like it applies to the overall step, but right now it's nested under step 1c. If you mean this to be a note for the entire step, you could try these two options instead:

1. Moving this note above the second level substeps
2. (Not sure this will work but) You could try adding `--` in between lines 35-36, and in between lines 91-92. This will tell asciibinder to treat everything in that `--` block as if it was its own discrete block of asciidoc text, and thus the `+` on line 92 might be interpreted as a continuation of the overall step and not step 1c. Not sure if it'll work though, maybe the `..` of the substeps won't be interpreted correctly

Up to you if either of these options interest you!

Author: Thanks for the suggestion @skopacz1 
I went for option 2 here and yes it worked

====
Each `ConfigMap` CR must be in the same namespace as the policy to be generated from the group `PolicyGenTemplate` CR.
====

. Commit the `ConfigMap` CRs in Git, and then push to the Git repository being monitored by the Argo CD application.

. Apply the hardware type and region labels to the clusters. 
The following command applies to a single cluster named `du-sno-1-zone-1` and the labels chosen are `"hardware-type": "hardware-type-1"` and `"group-du-sno-zone": "zone-1"`:
+
[source,terminal]
----

> oc patch managedclusters.cluster.open-cluster-management.io/du-sno-1-zone-1 --type merge -p '{"metadata":{"labels":{"hardware-type": "hardware-type-1", "group-du-sno-zone": "zone-1"}}}'

Reviewer: Add a $ at the beginning of the command here:
```suggestion
$ oc patch managedclusters.cluster.open-cluster-management.io/du-sno-1-zone-1 --type merge -p '{"metadata":{"labels":{"hardware-type": "hardware-type-1", "group-du-sno-zone": "zone-1"}}}'
```

.Procedure

> . Create three `ConfigMap` resources that describe the group and site configuration. 

Reviewer 2: ```suggestion
. Create three `ConfigMap` resources that contain the group and site configuration. 
```

+
.. Create `group-hardware-types-configmap`, a `ConfigMap` to hold the hardware specific configuration. For example:
+
[source,yaml]
----
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: group-hardware-types-configmap
  namespace: ztp-group
  annotations:
    argocd.argoproj.io/sync-options: Replace=true <1>
data:
  # SriovNetworkNodePolicy.yaml - general

>   dell-poweredge-xr12-sriov-node-policy-pfNames-1: "[\"ens5f0\"]"

Reviewer 2: We should probably change all occurrences of "dell-poweredge-xr12" to "hardware-type-1" to make it more generic.

  dell-poweredge-xr12-hugepages-count: "32"
----
<1> The `argocd.argoproj.io/sync-options` annotation is required only if the `ConfigMap` is larger than 1 MiB in size.

.. Create `group-zones-configmap`, a `ConfigMap` to hold the regional configuration. For example:
+
[source,yaml]
----
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: group-zones-configmap
  namespace: ztp-group
data:
  # ClusterLogForwarder.yaml
  zone-1-cluster-log-fwd-outputs: "[{\"type\":\"kafka\", \"name\":\"kafka-open\", \"url\":\"tcp://10.46.55.190:9092/test\"}]"
  zone-1-cluster-log-fwd-pipelines: "[{\"inputRefs\":[\"audit\", \"infrastructure\"], \"labels\": {\"label1\": \"test1\", \"label2\": \"test2\", \"label3\": \"test3\", \"label4\": \"test4\"}, \"name\": \"all-to-default\", \"outputRefs\": [\"kafka-open\"]}]"
----

.. Create `site-data-configmap`, a `ConfigMap` to hold the site specific configuration. For example:
+
[source,yaml]
----
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: site-data-configmap
  namespace: ztp-group
data:
  # SriovNetwork.yaml
  du-sno-1-zone-1-sriov-network-vlan-1: "140"
  du-sno-1-zone-1-sriov-network-vlan-2: "150"
----
+
[NOTE]
====

> Each `ConfigMap` must be in the same namespace with the policy that has the hub template substitution.

Reviewer 2: ```suggestion
Each `ConfigMap` must be in the same namespace with the policy that will be generated from the group PGT and is using hub templates.
```

====

. Commit the `ConfigMap` CRs in Git, and then push to the Git repository being monitored by the Argo CD application.

. Apply the hardware type and region labels to the clusters. 
The following command applies to a single cluster `named du-sno-1-zone-1` and the labels chosen are `"hardware-type": "dell-poweredge-xr12"` and `"group-du-sno-zone": "zone-1"`.
+
[source,terminal]
----

> oc patch managedclusters.cluster.open-cluster-management.io/du-sno-1-zone-1 --type merge -p '{"metadata":{"labels":{"hardware-type": "dell-poweredge-xr12", "group-du-sno-zone": "zone-1"}}}'

Reviewer 2: ```suggestion
oc patch managedclusters.cluster.open-cluster-management.io/du-sno-1-zone-1 --type merge -p '{"metadata":{"labels":{"hardware-type": "hardware-type-1", "group-du-sno-zone": "zone-1"}}}'
```

----

> . Create a group `PolicyGenTemplate` CR that uses templates to pull the required data from the `ConfigMap` objects. 

Reviewer 2: ```suggestion
. Create a group `PolicyGenTemplate` CR that uses templates to obtain the required data from the `ConfigMap` objects. 
```

This example `PolicyGenTemplate` CR configures logging, VLAN IDs, NICs and Performance Profile for the clusters that match the labels listed under `spec.bindingRules`:
+
[source,yaml]
----
apiVersion: ran.openshift.io/v1
kind: PolicyGenTemplate
metadata:
  name: group-du-sno-pgt
  namespace: ztp-group
spec:
  bindingRules:
    # These policies will correspond to all clusters with these labels
    group-du-sno-zone: "zone-1"

>     hardware-type: "dell-poweredge-xr12"

Reviewer 2: ```suggestion
    hardware-type: "hardware-type-1"
```

          pages:
            - size: '{{hub fromConfigMap "" "group-hardware-types-configmap" (printf "%s-hugepages-size" (index .ManagedClusterLabels "hardware-type")) hub}}'
              count: '{{hub fromConfigMap "" "group-hardware-types-configmap" (printf "%s-hugepages-count" (index .ManagedClusterLabels "hardware-type")) | toInt hub}}'
        realTimeKernel:
          enabled: true

    - fileName: SriovNetwork.yaml # wave 100
      policyName: "group-du-sno-sriov-policy"
      metadata:
        name: sriov-nw-du-fh
      spec:
        resourceName: du_fh
        vlan: '{{hub fromConfigMap "" "site-data-configmap" (printf "%s-sriov-network-vlan-1" .ManagedClusterName) | toInt hub}}'
    - fileName: SriovNetworkNodePolicy.yaml # wave 100
      policyName: "group-du-sno-sriov-policy"
      metadata:
        name: sriov-nnp-du-fh
      spec:
        deviceType: netdevice
        isRdma: false
        nicSelector:
          pfNames: '{{hub fromConfigMap "" "group-hardware-types-configmap" (printf "%s-sriov-node-policy-pfNames-1" (index .ManagedClusterLabels "hardware-type")) | toLiteral hub}}'
        numVfs: 8
        priority: 10
        resourceName: du_fh

>     - fileName: SriovNetwork.yaml # wave 100

Reviewer 2: Add blank line before.

      policyName: "group-du-sno-sriov-policy"
      metadata:
        name: sriov-nw-du-mh
      spec:
        resourceName: du_mh
        vlan: '{{hub fromConfigMap "" "site-data-configmap" (printf "%s-sriov-network-vlan-2" .ManagedClusterName) | toInt hub}}'

>     - fileName: SriovNetworkNodePolicy.yaml # wave 100

Reviewer 2: Add blank line before.

      policyName: "group-du-sno-sriov-policy"
      metadata:
        name: sriov-nw-du-fh
      spec:
        deviceType: netdevice
        isRdma: false
        nicSelector:
          pfNames: '{{hub fromConfigMap "" "group-hardware-types-configmap" (printf "%s-sriov-node-policy-pfNames-2" (index .ManagedClusterLabels "hardware-type")) | toLiteral hub}}'
        numVfs: 8
        priority: 10
        resourceName: du_fh
----
+
[NOTE]
====

> To retrieve site-specific configuration values, use the `ManagedClusterName` field. 

Reviewer 2: ```suggestion
To retrieve site-specific configuration values, use the `.ManagedClusterName` field. 
```

This is a template context value set to the name of the target managed cluster.

> To retrieve group-specific configuration, use the `ManagedClusterLabels` field. 

Reviewer 2: ```suggestion
To retrieve group-specific configuration, use the `.ManagedClusterLabels` field. 
```

This is a template context value set to the value of the managed cluster’s labels.
====

. Commit the site `PolicyGenTemplate` CR in Git and push to the Git repository that is monitored by the ArgoCD application.
+
[NOTE]
====
Subsequent changes to the referenced `ConfigMap` CR are not automatically synced to the applied policies. 
You need to manually sync the new `ConfigMap` changes to update existing `PolicyGenTemplate` CRs. See "Syncing new ConfigMap changes to existing PolicyGenTemplate CRs".

> You can use same `PolicyGenTemplate` CR for multiple clusters. 

Reviewer 2: ```suggestion
You can use the same `PolicyGenTemplate` CR for multiple clusters. 
```

> The only chnages you need to make are to the three `ConfigMap` objects that hold the configuration for each cluster and the labels of the managed clusters.

Reviewer 2: ```suggestion
In case there is a configuration change, the only modifications you need to make are to the `ConfigMap` objects that hold the configuration for each cluster and the labels of the managed clusters.
```

.Procedure

. Create three `ConfigMap` CRs that contain the group and site configuration. 
+
.. Create `group-hardware-types-configmap`, a `ConfigMap` to hold the hardware specific configuration. For example:
+
[source,yaml]
----
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: group-hardware-types-configmap
  namespace: ztp-group
  annotations:
    argocd.argoproj.io/sync-options: Replace=true <1>
data:

>   # SriovNetworkNodePolicy.yaml - general

Reviewer 2: ```suggestion
  # SriovNetworkNodePolicy.yaml
```

kind: ConfigMap
metadata:
  name: group-zones-configmap
  namespace: ztp-group
data:
  # ClusterLogForwarder.yaml
  zone-1-cluster-log-fwd-outputs: "[{\"type\":\"kafka\", \"name\":\"kafka-open\", \"url\":\"tcp://10.46.55.190:9092/test\"}]"
  zone-1-cluster-log-fwd-pipelines: "[{\"inputRefs\":[\"audit\", \"infrastructure\"], \"labels\": {\"label1\": \"test1\", \"label2\": \"test2\", \"label3\": \"test3\", \"label4\": \"test4\"}, \"name\": \"all-to-default\", \"outputRefs\": [\"kafka-open\"]}]"
----

.. Create `site-data-configmap`, a `ConfigMap` to hold the site specific configuration. For example:
+
[source,yaml]
----
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: site-data-configmap
  namespace: ztp-group
data:
  # SriovNetwork.yaml
  du-sno-1-zone-1-sriov-network-vlan-1: "140"
  du-sno-1-zone-1-sriov-network-vlan-2: "150"
----
+
[NOTE]
====
Each `ConfigMap` must be in the same namespace as the policy to be generated from the group `PolicyGenTemplate` CR that is using hub templates.
====

. Commit the `ConfigMap` CRs in Git, and then push to the Git repository being monitored by the Argo CD application.

. Apply the hardware type and region labels to the clusters. 

> The following command applies to a single cluster `named du-sno-1-zone-1` and the labels chosen are `"hardware-type": "hardware-type-1"` and `"group-du-sno-zone": "zone-1"`:

Reviewer 2: ```suggestion
The following command applies to a single cluster named `du-sno-1-zone-1` and the labels chosen are `"hardware-type": "hardware-type-1"` and `"group-du-sno-zone": "zone-1"`:
```

+
[source,terminal]
----
oc patch managedclusters.cluster.open-cluster-management.io/du-sno-1-zone-1 --type merge -p '{"metadata":{"labels":{"hardware-type": "hardware-type-1", "group-du-sno-zone": "zone-1"}}}'
----

> . Create a group `PolicyGenTemplate` CR that uses templates to obtain the required data from the `ConfigMap` objects. 

Reviewer 2: ```suggestion
. Create a group `PolicyGenTemplate` CR that uses hub templates to obtain the required data from the `ConfigMap` objects. 
```

          reserved: '{{hub fromConfigMap "" "group-hardware-types-configmap" (printf "%s-cpu-reserved" (index .ManagedClusterLabels "hardware-type")) hub}}'
        hugepages:
          defaultHugepagesSize: '{{hub fromConfigMap "" "group-hardware-types-configmap" (printf "%s-hugepages-default" (index .ManagedClusterLabels "hardware-type")) hub}}'
          pages:
            - size: '{{hub fromConfigMap "" "group-hardware-types-configmap" (printf "%s-hugepages-size" (index .ManagedClusterLabels "hardware-type")) hub}}'
              count: '{{hub fromConfigMap "" "group-hardware-types-configmap" (printf "%s-hugepages-count" (index .ManagedClusterLabels "hardware-type")) | toInt hub}}'
        realTimeKernel:
          enabled: true

    - fileName: SriovNetwork.yaml # wave 100
      policyName: "group-du-sno-sriov-policy"
      metadata:
        name: sriov-nw-du-fh
      spec:
        resourceName: du_fh
        vlan: '{{hub fromConfigMap "" "site-data-configmap" (printf "%s-sriov-network-vlan-1" .ManagedClusterName) | toInt hub}}'

>     - fileName: SriovNetworkNodePolicy.yaml # wave 100

Reviewer 2: Blank line before for better visibility.

      spec:
        deviceType: netdevice
        isRdma: false
        nicSelector:
          pfNames: '{{hub fromConfigMap "" "group-hardware-types-configmap" (printf "%s-sriov-node-policy-pfNames-2" (index .ManagedClusterLabels "hardware-type")) | toLiteral hub}}'
        numVfs: 8
        priority: 10
        resourceName: du_fh
----
+
[NOTE]
====
To retrieve site-specific configuration values, use the `.ManagedClusterName` field. 
This is a template context value set to the name of the target managed cluster.

To retrieve group-specific configuration, use the `.ManagedClusterLabels` field. 
This is a template context value set to the value of the managed cluster’s labels.
====

. Commit the site `PolicyGenTemplate` CR in Git and push to the Git repository that is monitored by the ArgoCD application.
+
[NOTE]
====
Subsequent changes to the referenced `ConfigMap` CR are not automatically synced to the applied policies. 
You need to manually sync the new `ConfigMap` changes to update existing `PolicyGenTemplate` CRs. See "Syncing new ConfigMap changes to existing PolicyGenTemplate CRs".

You can use the same `PolicyGenTemplate` CR for multiple clusters. 

> If there is a configuration change, then the only modifications you need to make are to the three `ConfigMap` objects that hold the configuration for each cluster and the labels of the managed clusters.

Reviewer 2: ```suggestion
If there is a configuration change, then the only modifications you need to make are to the `ConfigMap` objects that hold the configuration for each cluster and the labels of the managed clusters.
```
I wouldn't say `to the three ConfigMaps` because the change might have to be to just one ConfigMap or to 2 ConfigMaps. Similarly, the users might have a different number of ConfigMaps in their setup.

.Procedure

. Create three `ConfigMap` CRs that contain the group and site configuration. 
+

> .. Create `group-hardware-types-configmap`, a `ConfigMap` to hold the hardware specific configuration. For example:

Reviewer 3: ```suggestion
.. Create a `ConfigMap` CR named `group-hardware-types-configmap` to hold the hardware-specific configuration. For example:
```

+
[source,yaml]
----
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: group-hardware-types-configmap
  namespace: ztp-group
  annotations:
    argocd.argoproj.io/sync-options: Replace=true <1>
data:
  # SriovNetworkNodePolicy.yaml
  hardware-type-1-sriov-node-policy-pfNames-1: "[\"ens5f0\"]"
  hardware-type-1-sriov-node-policy-pfNames-2: "[\"ens7f0\"]"
  # PerformanceProfile.yaml
  hardware-type-1-cpu-isolated: "2-31,34-63"
  hardware-type-1-cpu-reserved: "0-1,32-33"
  hardware-type-1-hugepages-default: "1G"
  hardware-type-1-hugepages-size: "1G"
  hardware-type-1-hugepages-count: "32"
----
<1> The `argocd.argoproj.io/sync-options` annotation is required only if the `ConfigMap` is larger than 1 MiB in size.

> .. Create `group-zones-configmap`, a `ConfigMap` to hold the regional configuration. For example:

Reviewer 3: ```suggestion
.. Create a `ConfigMap` CR named `group-zones-configmap` to hold the regional configuration. For example:
```

+
[source,yaml]
----
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: group-zones-configmap
  namespace: ztp-group
data:
  # ClusterLogForwarder.yaml
  zone-1-cluster-log-fwd-outputs: "[{\"type\":\"kafka\", \"name\":\"kafka-open\", \"url\":\"tcp://10.46.55.190:9092/test\"}]"
  zone-1-cluster-log-fwd-pipelines: "[{\"inputRefs\":[\"audit\", \"infrastructure\"], \"labels\": {\"label1\": \"test1\", \"label2\": \"test2\", \"label3\": \"test3\", \"label4\": \"test4\"}, \"name\": \"all-to-default\", \"outputRefs\": [\"kafka-open\"]}]"
----

> .. Create `site-data-configmap`, a `ConfigMap` to hold the site specific configuration. For example:

Reviewer 3: ```suggestion
.. Create `site-data-configmap`, a `ConfigMap` to hold the site specific configuration. For example:

.. Create a `ConfigMap` CR named `site-data-configmap` to hold the site-specific configuration. For example:
```

+
[source,yaml]
----
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: site-data-configmap
  namespace: ztp-group
data:
  # SriovNetwork.yaml
  du-sno-1-zone-1-sriov-network-vlan-1: "140"
  du-sno-1-zone-1-sriov-network-vlan-2: "150"
----
+
[NOTE]
====

> Each `ConfigMap` must be in the same namespace as the policy to be generated from the group `PolicyGenTemplate` CR that is using hub templates.

Reviewer 3: ```suggestion
Each `ConfigMap` CR must be in the same namespace as the policy to be generated from the group `PolicyGenTemplate` CR.
```

====

. Commit the `ConfigMap` CRs in Git, and then push to the Git repository being monitored by the Argo CD application.

. Apply the hardware type and region labels to the clusters. 

> The following command applies to a single cluster named `du-sno-1-zone-1` and the labels chosen are `"hardware-type": "hardware-type-1"` and `"group-du-sno-zone": "zone-1"`:

Reviewer 3: Is this correct? Wouldn't this command patch any resource with the label `managedclusters.cluster.open-cluster-management.io/du-sno-1-zone-1`?

Reviewer 2: This command is correct. It will patch the labels for the `du-sno-1-zone-1` managed cluster.

        numVfs: 8
        priority: 10
        resourceName: du_fh

    - fileName: SriovNetwork.yaml # wave 100
      policyName: "group-du-sno-sriov-policy"
      metadata:
        name: sriov-nw-du-mh
      spec:
        resourceName: du_mh
        vlan: '{{hub fromConfigMap "" "site-data-configmap" (printf "%s-sriov-network-vlan-2" .ManagedClusterName) | toInt hub}}'

    - fileName: SriovNetworkNodePolicy.yaml # wave 100
      policyName: "group-du-sno-sriov-policy"
      metadata:
        name: sriov-nw-du-fh
      spec:
        deviceType: netdevice
        isRdma: false
        nicSelector:
          pfNames: '{{hub fromConfigMap "" "group-hardware-types-configmap" (printf "%s-sriov-node-policy-pfNames-2" (index .ManagedClusterLabels "hardware-type")) | toLiteral hub}}'
        numVfs: 8
        priority: 10
        resourceName: du_fh
----
+
[NOTE]
====
To retrieve site-specific configuration values, use the `.ManagedClusterName` field. 
This is a template context value set to the name of the target managed cluster.

To retrieve group-specific configuration, use the `.ManagedClusterLabels` field. 

> This is a template context value set to the value of the managed cluster’s labels.

Reviewer 3: ```suggestion
This is a template context value set to the value of the managed cluster's labels.
```
Use the `'` (apostrophe) for possessives.
